src/mame/jaleco/ms1_gatearray.h

```diff
@@ -0,0 +1,97 @@
+// license:BSD-3-Clause
+// copyright-holders:Angelo Salese, David Haywood
+
+#ifndef MAME_JALECO_MS1_GATEARRAY_H
+#define MAME_JALECO_MS1_GATEARRAY_H
+
+#pragma once
+
+class megasys1_gatearray_device : public device_t
+{
+public:
+	template <typename T> void set_cpu_tag(T &&tag) { m_cpu.set_tag(std::forward<T>(tag)); }
+	void set_rom(u8* rom, u32 size) { m_rom = rom; m_romsize = size; }
+
+protected:
+	megasys1_gatearray_device(
+			machine_config const &mconfig,
+			device_type type,
+			char const *tag,
+			device_t *owner,
+			u32 clock);
+
+	virtual void device_start() override;
+	virtual void device_reset() override;
+
+	void install_gatearray_overlay();
+	u16 gatearray_r(offs_t offset, u16 mem_mask = ~0);
+	void gatearray_w(offs_t offset, u16 data, u16 mem_mask = ~0);
+
+	int m_gatearray_hs = 0;
+	u16 m_gatearray_hs_ram[0x8]{};
+
+	const u16* m_gatearray_seq = nullptr;
+
+	required_device<cpu_device> m_cpu;
+	u8* m_rom;
+	u32 m_romsize;
```

Comment:
Please use a `required_memory_region` for this, so it can be configured at machine config time, similarly to the CPU.  Then you can do the decoding on device start instead of having to do it on reset and keep track of whether you’ve done it.